watchfromcli.1
.TH watchfromcli 1 "7 November 2017" Linux "User Manuals"
.\"
.\" NAME ----------------------------------------------------------------------
.\"
.SH NAME
watchfromcli \- a shell wrapper for mpv/MPlayer to run videos easy via CLI.
.\"
.\" SYNOPSIS ------------------------------------------------------------------
.\"
.SH SYNOPSIS
.TP 4
.IP "Simplest form"
.B watchfromcli
[\fIoptional arguments\fP] \fB-d\fP \fIbasepath  keyword\fP
.IP "To start watching cycle"
.br
.B watchfromcli
[\fIoptional arguments\fP] \fB-c -d\fP \fIbasepath  keyword\fP
.IP "To resume watching cycle"
.br
.B watchfromcli
[\fIoptional arguments\fP] \fB-[r|R]\fP  [\fIkeyword\fP]
.IP "Check for updates"
.br
.B watchfromcli -u
.\"
.\" DESCRIPTION ---------------------------------------------------------------
.\"
.SH DESCRIPTION
\fIwatchfromcli\fP is written to simplify access to video files and play videos easily. It allows playing single files, DVD/BD folders and folders with episodes of TV series. For the former and the latter ones it can also find external subtitles and tracks, applying heuristic algorithms to include files with names that may strongly differ from the names of corresponding video files, and that frees you from the job of picking appropriate subtitles manually.

This script can also help you deal with organization of taken screenshots, placing them right into the folder named after video and compressing them after you stop watching it.

For the folder with episodes \fIwatchfromcli\fP can track your watching progress, allowing to run episodes one after another without creating a playlist, but remembering the episode number you quit watching today, so you could easily resume watching the folder later.

\[char46]\[char46]\[char46]and many other accompanying things that make life even lazier.
.\"
.\" LIST OF ARGUMENTS ---------------------------------------------------------
.\"
.SH LIST OF ARGUMENTS
.SS "Required arguments"
.TP
.B -d, --basedir, --basepath <\fIbasepath\fP>
Main folder to start search in.
.TP
.I keyword
.br
A string to search in file and folder names. It is meant to handle simple strings in 98.9% of cases, but it also should support BRE.
.br
An example of strings that can be passed as keyword:
.RS 10
-
.I hachirota
-- plain alphanumeric;
.br
-
.I 'abc|d ef'
-- a pattern which would find anything containing \fBabc\fP or \fBd ef\fP;
.br
-
.I 'khihihi.*| ahem | nah'
-- possible, but unnecessary;
.br
-
.I 'Gzh[f-z]+[0-9]{2} | oooh~'
-- theoretically possible <s>in the future.</s>
.RE
.TP
.B -c, --run-in-cycle
Starts watching cycle in a found folder.
.P
.B -r, --resume
.br
.B -R, --resume-from-previous
.br
.RS 7
Resumes watching cycle. When the \fIkeyword\fP is given (the keyword is an argument by itself), it resumes session started with the specified \fIkeyword\fP, and the last session otherwise. Sessions are stored in a journal which size is limited (see also \fBFILES\fP section and descriptions to \fB--no-journal\fP and \fB--journal-max-size\fP options).

.br
\fB-r\fP|\fB--resume\fP will play the next episode, if the last one was watched till the end, or the previous one, if watching was interrupted (player was closed before the end of file has been reached). \fB-R\fP|\fB--resume-previous\fP is to select one episode back from what usual \fB--resume\fP would take. It may be helpful in case if you accidentally rewind the episode till the end -- to avoid starting a new watching cycle to open that previous episode again.
.RE
.\"
.\" OPTIONAL ARGUMENTS --------------------------------------------------------
.\"
.SS "Optional arguments"
.P
.B -a, --match-all
.br
.B -n, --match-number
.br
.RS 7
When the script is looking for external subtitles and tracks, it always picks those matching the video file name, but depending on the mode the script running in, it may add some another. Thus, in the mode when it plays a sequence of \fIepisodes\fP, the search for external subtitles and tracks is also done for those having both \fIkeyword\fP and \fIepisode number\fP in their names. And when it plays a \fIsingle\fP file, it also searches for those matching \fIkeyword\fP. \fB--match-number\fP allows external files having only the \fIepisode number\fP to be included to the list of found subtitles, while \fB--match-all\fP in addition to that includes those having only the \fIkeyword\fP in their names.
.RE
.TP
.B --bashrc[=\fIpath to file\fP]
.br
Source ~/.bashrc or the file specified as an argument. This is to make aliases available for \fB--mplayer-command\fP.
.TP
.B -C, --no-color
Forbid colored output, if possible.
.TP
.B --compat <mplayer|mplayer2|mpv-03x|mpv-025x>
Set compatibility mode for keys passing to \fB--mplayer-command\fP. This affects DVD/BD playing, adding external subtitles and tracks, which is involving building options for the player command. The order is time-based. mpv-03x means 0.3.x and mpv-025x means 0.25.x. Zehkul’s convert_script.lua doesn’t work with >=mpv-0.25.
.br
The script is written to conform with the syntax of the latest \fBmpv\fP build (i.e. from git), so there's no need in compatibility modes if you use the upstream.
Use "mplayer" for the original MPlayer, "mplayer2" for mplayer2 and mpv-03x for mpv-0.3.x or older.
.TP
.B -e
Enable heuristic methods for finding episode numbers in file names. Specifying this option increases the default \fIheuristics level\fP, which initially equals to 0. This option is cumulative and since the highest \fIheuristics level\fP is 2, \fB-ee\fP is enough to bump it to the maximum.
Usually, there's no need to enable heuristics at all, if the files named properly and those names not differ much.
.RS 10
.B \-e
is useful when there are two or more kinds of names;
.br
.B \-ee
is only needed, if the sequence is comprised of several files which names highly differ. What was former 2nd level now is included in 1st, and the current second level is under development now.
.RS -3
See \fBUsing heuristics\fP subsection for details.
.RE
.RE
.TP
.B -f
Stands for "fixed string", tells \fIwatchfromcli\fP to treat the \fIkeyword\fP as a literal. Don't count on this switch much, it affects the code in merely two places, and fixed strings won't be possible in the future, because there are unavoidable expressions where the \fIkeyword\fP is used in patterns, so if this switch will develop, it would be escaping, not a real fixed string.
.TP
.B --group-indicator <\fIchars\fP>
If for some reason, the terminal can't display ASCII graphic characters and the bullet, you can specify another set via this option. \fIchars\fP must be a string of exactly four characters, the first for the beginning of a group, the second for middle elements of the group, the third for the ending and the fourth for lone ones.
.TP
.B --heuristics-level <0..2>
Sets heuristics level directly to a specified number. The second level is under development and is currently disabled.
.TP
.B -H <\fIpattern\fP>
Calls \fBman watchfromcli\fP with 'pager' option, specifying \fBless\fP as the pager and passing \fIpattern\fP to it, allowing to open this man page on a specified word, e.g. passing \fB-H '^EXAM'\fP will open  this man page on \fBEXAMPLES OF USE\fP section.
.TP
.B -h, --help
Shows this man page.
.TP
.B -I, --ignore-disks
Treat a folder containing DVD of BD stuff like a folder with episodes instead of playing it like an actual DVD or BD (See also \fBPlaying DVD and BD\fP subsection).
.TP
.B --interval <\fB0..9999999\fP>
Interval between playing episodes, in seconds. It is always set to 0 if mpv must play DVD or BD contents as episodes. The default is 3 seconds.
.TP
.B --ionice-opts [\fIoption list\fP]
Run mpv with ionice command. This is intended to give mpv higher I/O priority, when several tasks use the same drive. If \foption list\fP is omitted, watchfromcli assumes "-c best-effort -n0" - the best we can give without root access.
.TP
.B -J, --no-journal
Do not use \fIjournal\fP to store session data. Resuming the cycle won't work, but you can gain a bit of privacy instead.
.TP
.B --journal-max-size <\fIvalue\fP>
The maximum size the \fIjournal\fP will be truncated to, in bytes. Additional suffixes K, M and G are recognized in the \fIvalue\fP to represent KiB, MiB and GiB, e.g. "1025" is 1025 bytes, "1K" is one kibibyte (1*1024 = 2^10 bytes). The default size is 64K.
.TP
.B --jpeg-compression[=0..100]
Use conversion to JPEG after running \fBpngcrush\fP (if present) on screenshots taken after closing watch session. The optional parameter defines quality, the default is 92. Use this only if you have MPlayer, mpv can save screenshots to JPEG itself, see \fBman mpv\fP to know how to set up its configuration file.
.TP
.B -L, --limit-watching-to <0..9999>
Long-awaited option to limit the nnumber of episodes that would be played in a sequence. I.e. if you're watching something, say, from 26 episodes and only have time to watch three, pass \fB-L3\fP along with the \fB-c\fP, \fB-r\fP or \fB-R\fP option, and watchfromcli will quit after playing the episode that stands after 3 positions in the list (so one can safely do \fB-R\fP between episodes), e.g. if the watching has started on 14th episode, watchfromcli will quit when 16th ends.
.TP
.B -l, --loop
Loop the watching cycle.
.TP
.B --last-ep
When watching goes in a \fIcycle\fP, prints the last shown episode number in big ASCII-art numbers, so it would be easier to remember. It sets three internal variables to default values:
.RS 10
- \fBcommand\fP - command line calling the program to output in ASCII-art (consider installing \fBfiglet\fP or \fBtoilet\fP for this purpose), default value is "figlet -t -f clb6x10 -c" or "cat" if figlet is not installed yet;
.br
- \fBformat\fP is the string to print. By default consists of only "%n", which is substituted with the number of the last shown episode. "\\n" and some other backslash combinations are also recognized, since the output is passed through \fBecho -e\fP;
.br
- \fBshow-after\fP - defines the time when to output the number. It may be printed after the player stops (and before screenshots processing), after processing, or in both times (this is the default), in this case the number is printed the second time only if there were screenshots that required processing.
.RS -3
This option isn't necessary to show the last episode number, it only sets default values, which you can provide separately. The actual trigger is any correct value to \fBshow-after\fP (see below).
.RE
.RE
.P
.B --last-ep-command <\fIcommand\fP>
.br
.B --last-ep-format <\fIformat\fP>
.br
.B --last-ep-show-after <player|screenshots|both>
.br
.RS 7
These options are a replacement for \fB--last-ep\fP, for those who likes to customize. For \fIcommand\fP and \fIformat\fP description, see \fB--last-ep\fP.
.RE
.TP
.B -M, --mplayer-command <\fIcommand\fP>
The \fIcommand\fP to run the player. May be an executable or an \fIalias\fP. \fBmpv\fP is used by default.
.TP
.B -m, --mplayer-opts <\fIoptions\fP>
\fIoptions\fP are passed to the \fImplayer command\fP after found subtitles and tracks (if there are any).
.br
P.S. mpv's --profile options are cumulative, so one can easily pass --profile=$HOSTNAME and --profile=hdmi for example.
.TP
.B --last-item-mark <\fImark\fP>
When mpv finishes playing the last item, if printing the episode number is enabled, watchfromcli can print also a mark, be it a dot "." or a word on the next line "\\nEND". This is to see if the cycle has actually ended in case there were a bunch of screenshots. The thing is, after the last number and \fBparallel\fP header are printed, there is a small pause before output from parallel starts, so you look at the screen for several seconds for nothing. Having a mark tells when the watch cycle is finished, so you can move on to something else while parallel works.
.TP
.B --my-increment <\fIchar\fP>
.RS 0
.B --my-decrement <\fIchar\fP>
.RS 7
For characters to increment and decrement the number in the menu, if you want a pair of keys to supplement arrow keys on the other edge of keyboard. \fIchar\fP may be specified like a letter (k), an octal number ($'\\000') or as an escape sequence ($'\\e[D'). If heuristics is enabled (i.e. its level >0), then TAB, "h" and "H" will be reserved for their mode changing functions.
.br
In bash you can print escape sequence of a character generated by a key with \fBC-v <key>\fP.
.RE
.TP
.B -N, --dvd-bd-nav
Pass navigation variant of the protocol when playing a disk, i.e. dvdnav:// and bdnav:// instead of dvd:// or bd://. bdnav:// is only supported by \fBmpv\fP.
.TP
.B --no-hints
Do not show hints built in the interface (for experienced users).
.TP
.B --not-epnumbers <\fIpattern\fP>
For files that don't fall in any sequence, it is harder to define, where the episode number may be, and whether it is actually present there. In general case any number found in the filename may be the episode number, but watchfromcli has, however, an array of patterns which, if met, cut off these patterns and all the rest till the end of the filename. These patterns include strings like 720p, x264 etc. The full list is defined in the \fBNOT_EPNUMBERS\fP variable in watchfromcli. Some of the strings are \fIpatterns\fP, so you can add elements which are \fIpatterns\fP, too (useful commits are appreciated). Pattern is a bash pattern with extended glob matching (see \fBPattern Matching\fP section in the bash manual page). One pattern per option call allowed.
.TP
.B --remember-sub-and-audio-delay
Should be self-explanatory. But, for this to work one must set up certain mpv keybinding. See the details under \fBRemembering subtitles and audio delay between episodes\fP subsection.
.TP
.B -s, --subfolders <\fIpattern list\fP>
Prospective subfolders that are likely expected between a folder found in \fIbasepath\fP and files matching the \fIkeyword\fP. Usually they are subfolders that divide seasons or spans of episodes and named like "0-100", so putting "season -" in the \fIpattern list\fP will order to perform recursive search into the folder found in \fIbasepath\fP appending directories that do not contain \fIkeyword\fP but do contain word "season" and/or dash "-" and look for video at the end of the new path.
\fB%keyword\fP in the pattern list will be substituted with actual \fIkeyword\fP passed via command line.
.TP
.B -S, --screenshot-dir <\fIscreens path\fP>
If set, makes the script look for a folder *inside* the \fIscreens path\fP containing \fIkeyword\fP in its name. The script will change directory to it, so all the taken screenshots would be put in there. It will also try to compress fresh screenshots of PNG format with \fBpngcrush\fP if it will be found. For subsequent conversion to JPEG see \fB--jpeg-compression\fP.
There’s no need to use this switch to make mpv use just ONE certain folder, that is not $PWD, for any screenshots you take. Instead, consider adding a \fBcd\fP command to the alias before calling \fBwatchfromcli\fP. The player should use it for storing screenshots.
.TP
.B --screenshot-dir-skel <\fIdirlist\fP>
The argument is a list of comma separated directories to be created under the folder specified via \fB--screenshot-dir\fP, e.g. "art,misc".
.TP
.B -t, --taskset-opts <\fICPU list\fP>
Passes \fICPU list\fP to \fBtaskset\fP, \fICPU list\fP may be a number, a span or a list, e.g. "0", "1-3", "0,2-4,7". It's called "taskset-opts" but actually takes only CPU list, so "-c" is not needed. Name is for conformance with --ionice-opts.
.P
.B -u
.br
.B --check-for-update[=0..n]
.br
.RS 7
By default, watchfromcli will check each 21 day if there is a new release and report the status. The check can be initiated by both \fB-u\fP and \fB--check-for-update\fP without arguments and watchfromcli will exit after the check, if it was specified with \fB-u\fP. If the argument to the long option is a number and is greater than 1, then it defines the pause in days between checks. The value of 0 suppresses any kind of automatic checks and concomitant messages. Immediate check can still take effect.
.RE
.TP
.B -v, --version
Prints version and legal information.
.\"
.\" FILES ---------------------------------------------------------------------
.\"
.SH FILES
\fB~/.watchfromcli/journal\fP - stores session data.
.\"
.\" EXAMPLES OF USE -----------------------------------------------------------
.\"
.SH EXAMPLES OF USE

First of all, everyone has a directory with video files,
.I /home/video/
for example. In there can be single files (movies) or folders containing episodes, Blu-ray or DVD stuff. This is what must be passed to the \fB-d\fP option, which stands for \fIbasedir\fP or \fIbasepath\fP. To find something in it, a \fIkeyword\fP must also be passed. In this simlpiest case the script would
search for single files as well as folders inside of the \fIbasepath\fP, which names contain the specified \fIkeyword\fP. For example, if command line would look like that
.nf
	$ watchfromcli -d /home/video clash
.fi
then items in \fBbold\fP will be shown to choose one of them to continue:
.nf
	/home/video/\fBclash.mkv\fP                           # video file match
	/home/video/\fBthose ones clashing hardly\fP/01.mkv   # folder match
	                                      /02.mkv
	/home/video/clashing_sounds_10_hours.ogg        # not a video
	/home/videos/free clash.mp4                     # not in \fIbasepath\fP
.fi
If the chosen one was a folder, then search goes further in its contents. Here an additional option, \fB--subfolders\fP may have an effect - if there are folders under the chosen directory, their names will be appended to the resulting path to video files as long as they match any of the patterns in \fIpattern list\fP (see above). Let's assume the chosen directory was
.nf
	/home/video/\fBClashes in the morning\fP
.fi
which, in its turn, has the following structure
.nf
	./Season 1/01.mkv
	          /02.mkv
	          /...
	./Season 2/01.mkv
	          /...
.fi
then passing \fB-s 'season'\fP will perform recursive search in the current directory, appending inner folders containing 'season' (case-insensitive) in their names to path, so .mkv files could get to the list of episodes. See also description to the \fB--subfolders\fP option above.
.\" -----------------------------------------------------------------------SS--
.SS Using heuristics
Sometimes the file names in the destination folder do not have a common template, making the task to guess the correct order of the files harder for the script. To improve ordering results, there were introduced levels of applied heuristics. It's only two of them: 1 and 2. Zero means heuristics is disabled and ordering is performed by calling the \fBsort\fP utility. For the vast majority of cases that would be enough, also this is the fastest method.
.br
But this does not cover the rest of the cases, when the list would be built in the wrong order. And to deal with it, the first level of heuristics breaks up each file name from the list to patterns residing to the left and to the right of presumptive episode numbers, increments that number, and, if such file will be found in the list, it combines all the file names having numbers in that place in a group. The largest group will be at the top. Second level, in addition to the before mentioned, does check all numbers in groups as numbers (and not as symbolic strings), resorting matches inside groups again. This takes time and may lead to some delay on CPUs that are not that powerful (old Pentiums, Atoms, ARMs etc). But this helps to align the list of episodes in the most correct order, and, in combination with file name groups, in the most desired way (if there is a season and a number of OVAs in the folder, 12 openings, or episode numbers go like "1 2... 10 11" instead of "01 02... 10 11" which makes usual \fBsort\fP fail). It's not recommended to enable 2nd level on a default basis, like putting it in the alias, cause the higher is the level, the slower the script works. It is always available in one key reach.
.br

Heuristics can be changed in the runtime by pressing \fBh\fP and \fBH\fP keys to lower the level and to rise it, respectively. Setting the level with \fB-e\fP or \fB-ee\fP option on a default basis may enable to avoid manual switching. Here's the reference table of heuristic levels
.nf
.B "+-----+---------------------+----------------------------------------+"
\fB|\fPlevel\fB|\fPcommand needed to set\fB|\fP               description              \fB|\fP
.B "+-----+---------------------+----------------------------------------+"
\fB|\fP  0  \fB|\fP          -          \fB|\fP Heuristics disabled. Ordering is done  \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP via `sort` utility. This is the fast-  \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP est method.                            \fB|\fP
.B "+-----+---------------------+----------------------------------------+"
\fB|\fP  1  \fB|\fP -e                  \fB|\fP Each file name is split up to parts to \fB|\fP
\fB|\fP     \fB|\fP --heuristics-level 1\fB|\fP find numbers. Other parts combined in  \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP patterns, which in their turn are com- \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP bined in groups, items in groups are   \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP sorted by numbers and the largest      \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP grroup is placed at the top. Allows    \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP to define episode number more precise- \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP ly. Works a bit slower.                \fB|\fP
.B "+-----+---------------------+----------------------------------------+"
\fB|\fP  2  \fB|\fP --heuristics-level 2\fB|\fP Above what is done on the 1st level,   \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP watchfromcli will try to build the list    \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP of episodes, starting with complemen-  \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP ting the first group. If it finds      \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP a gap in episode numbers, it will try  \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP to find groups and single filenames,   \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP that would fit entirely, and do the    \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP corresponding rearrangements. After    \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP reaching the end of the group it will  \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP seek for a group or a single filename  \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP to continue the sequence of episode    \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP numbers it has found. For single files \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP it will check the numbers that are     \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP left after applying NOT_EPNUMBERS pat- \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP terns, and fill the gaps in the new    \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP group, and so forth.                   \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP Missing 1..n numbers for the 1st taken \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP group is also considered a gap.        \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP At current stage it is implemented     \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP at 97 %, the work reuired rewriting    \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP the heuristics almost entirely, and    \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP the fruits of this work are manual     \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP rearrangement and automated queue for  \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP rearrangements with prediction abili-  \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP ty that enables to perform many rear-  \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP rangements on indices and do the actu- \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP al rearrangement on real data at once  \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP to the resulting state. I hope         \fB|\fP
\fB|\fP     \fB|\fP                     \fB|\fP I could finish it sometimes.           \fB|\fP
.B "+-----+---------------------+----------------------------------------+"
.fi
.\" -----------------------------------------------------------------------SS--
.SS Using manual rearrangement
As a part of rewriting heu algorithms, there was implemented the possibility for the use to rearrange the items in the list. The syntax is simple: \fIfirst_item\fP[-\fIlast_item\fP]>\fIdestination_line\fP, e.g. \fB5-8>12\fP, \fB4>10\fP, they may also be combined via \fB,\fP (comma) like \fB1-2>12,9-10>4\fP. Numbers must represent lines as they currently seen on the screen. \fB1-2>12,4-5>2\fP will place elements from lines 1 and 2 to 12th line, and what is now seen on the lines fourth to fifth will be placed at the 12th line, too. Humans wouldn't need to do that.

.br
Manual rearrangement is supposed to be an ultimate fix, that user does when the automatic arrangement can't be brought to a satisfying view, and hence no further automatic arrangements are possible. The effect of manual rearrangement is dropped when switching between heuristics levels (in case it was made by a mistake).
.\" -----------------------------------------------------------------------SS--
.SS Playing DVD and BD

watchfromcli.sh
#! /usr/bin/env bash

#  watchfromcli.sh
#  A shell wrapper for mpv to launch videos easy via CLI.
#  © deterenkelt 2013–2019

 # This program is free software; you can redistribute it and/or modify it
#  under the terms of the GNU General Public License as published
#  by the Free Software Foundation; either version 3 of the License,
#  or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but without any warranty; without even the implied warranty
#  of merchantability or fitness for a particular purpose.
#  See the GNU General Public License for more details.


# Requires
# GNU sed >= 4.2.1 (started developing with it).
# GNU grep >= 2.9 (started developing with it).
# GNU bash >= 4.4 (strongly).
# mimetype >= 0.28
# util-linux >= 2.20 (for getopt that is required, and taskset
#        which may be of use, but is optional).
# mplayer, mplayer2 or mpv.
#
# Works better with
# GNU parallel – to compress screenshots faster using all cores available
#       (or those available after restricting to those specified
#        to the -t or --taskset option).
# figlet – to draw last seen episode number with big ASCII art numbers.
# pngcrush – helps to reduce PNG image size, if you prefer it over JPEG.
#       (players tend to save PNG in an unoptimized format, which makes
#        screenshots very large. pngcrush recompresses them without quality
#        loss).
# pngtopam and cjpeg – are only needed for converting screenshots from PNG
#       (if you, for some reason use MPlayer, that can only save them to PNG)
#        to JPEG by the usage of --jpeg-compression. pngtopam is usually found
#        in the netpbm package and cjpeg in libjpeg-turbo.
# inotifywait, ps and pkill – for SUB_DELAY, is of use only with mpv.
#        The first belongs to intotify-tools and the latter – to procps package.


# extglob for the sake of it, expand_aliases to make aliases available for
#   MPLAYER_COMMAND
shopt -s extglob expand_aliases
set -feEuT

BAHELITE_CHERRYPICK_MODULES=(
	error_handling
	logging
	misc
)
. "$(dirname "$(realpath --logical "$0")")/lib/bahelite/bahelite.sh"
prepare_cachedir
start_log

show_help() {
	cat <<-"EOF"
	Simplest form:
	    watch.sh [optional arguments] -d <basepath>  <keyword>

	To start watching cycle:
	    watch.sh [optional arguments] -c -d <basepath>  <keyword>

	To resume watching cycle:
	    watch.sh [optional arguments] <-r|-R>  [keyword]

	Open wiki <link>
	Send issues to Github <link>
	EOF
}


show_version() {
	cat <<-EOF
	watchfromcli.sh $VERSION
	© deterenkelt 2013–2019.
	License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
	This is free software: you are free to change and redistribute it.
	There is NO WARRANTY, to the extent permitted by law.
	EOF
}


cd "$PWD" 2>/dev/null || err 'Cannot run – current directory is removed.'



# Variables typed in caps are either
# - bash built-ins;
# - those ones that set parameters from the options passed through the command line;
# - used between functions;
# - or are important for maintaining the watching cycle between runs.

# 2nd level is 97% done, but I haven’t time to finish it while I was at it,
#   so the rest of the tests will require to dig up in all that again.
# MAX_HEURISTICS_LEVEL=2
MAX_HEURISTICS_LEVEL=1
HEURISTICS_LEVEL=0

VERSION="20190223"

JOURNAL=~/.watch.sh/journal
JOURNAL_MAX_SIZE="64K" # w/o suffix for bytes, K for KiB, M for MiB et al.
JOURNAL_MINVER='20180409'

[ -d ~/.watch.sh ] || {
	mkdir -m755 ~/.watch.sh/ >/dev/null \
		|| err "Couldn’t create directory ~/.watch.sh."
}



GROUP_INDICATOR='┌│└⋅' # upper part/middle part/lower part/single

# If any of these patterns is met in a filename, which episode number
#   can’t be guessed, then it and all numbers met further can’t be
#   taken as a presumed episode number.
# Patterns are to be given to bash with enabled extglob. To be precise,
#   they’ll be used in an expression like this:
#   "${var%${NOT_EPNUMBERS[i]}*}"
NOT_EPNUMBERS=("240p" "360p" "480p" "720p" "1280??(?)?(?)720" "1080p" "1920??(?)?(?)1080" "@(h|H|h.|H.|x)264" "10?bit")

NO_AUTOSUB='--sub-auto=no'

# DEBUG MODE
# No function aggregator for debug messages because test [ -v D ]
#   is faster than function call (there would be lots of them).
# Also [ -v D ] is more visually distinguishable. descartes.jpg
[ -v D ] && {
	DEBUG_DIR="$HOME/.watch.sh/debug"
	[ -d "$DEBUG_DIR" ] && rm -rf "$DEBUG_DIR"
	mkdir -m755 "$DEBUG_DIR" >/dev/null || err "Couldn’t create directory ‘$DEBUG_DIR’."
	for i in "$0" "$@"; do echo "\"$i\"" >>"$DEBUG_DIR/cmdline"; done
	vars="`set -o posix; set`"
}

set_libdir
#  Escaping for pattern and replacement.
. "$LIBDIR/sed_helpers.sh"
#  To be removed, when the entire code is reviewed and remade for Bahelite.
. "$LIBDIR/misc_helpers.sh"
set_modulesdir
set +f
for module in "$MODULESDIR"/watchfromcli_*.sh ; do
	. "$module" || err "Couldn’t source module $module."
done
set -f


is_this_the_last_item() {
	# If [ $MODE = bd ] then VIDEO* vars will be empty
	[    ${VIDEO_NUMBER:- 1} -eq ${VIDEOFILES_COUNT:- -1} \
      -o ${VIDEO_NUMBER:- 1} -eq ${EXIT_AFTER_THIS_EPISODE:- -1} ]
}

print_last_shown_episode_number() {
	echo
#	declare -p EP_NUMBERS
	local ep_number=${EP_NUMBERS[VIDEO_NUMBER-1]##*(0)}
	is_this_the_last_item && local last_item_finishing_mark=$LAST_ITEM_MARK
	$LAST_EP_NUMBER_PRINTING_COMMAND < \
		<( echo -e "${LAST_EP_NUMBER_PRINTING_FORMAT//%n/$ep_number}${last_item_finishing_mark:-}" )
}



                   #  Main algorithm starts here  #

process_args "${ARGS[@]}"
if [ -v RESUME ]; then
	import_session_data
else
	do_initial_search
fi
set_screenshot_subdir
# Exit trap should be here – after all the necessary data are collected or
#   imported. There is no point in altering the journal on exit, if user
#   declined to start watching something halfway.
on_exit() { export_session_data || exit $?; }
# A good place to check the $MODE.
until [ -v STOP ]; do
	watch || exit $?
	[ $MODE = episodes ] \
		&& [ "$LAST_EP_NUMBER_SHOW_AFTER" = player \
		  -o "$LAST_EP_NUMBER_SHOW_AFTER" = both   ] \
		&& print_last_shown_episode_number
	[ -v STOP ] || {
		[ -v RUN_IN_CYCLE ] && {
			is_this_the_last_item && {
				[ -v LOOP ] || break
				VIDEO_NUMBER=1
			}
			echo -en "Press $g<Space>$s to stop > "
			read -n1 -t ${INTERVAL:-3}
			# Avoiding [C to get to mpv’s input in case right arrow
			#   is still hold. If mpv reads them, it’ll try to interpret
			#   them as commands, and [ is usually bound to decrease speed
			#   by multiplying it on 0.9.
			[ "$REPLY" = $'\e' ] && read -sn2 rest && REPLY+="$rest"
			echo
			[ "$REPLY" = ' ' ] && break || IT_IS_NEXT_ITERATION=t
		}|| break
	}
done

# screenshots_postprocessing && [ $MODE = episodes ] \
	# && [    "$LAST_EP_NUMBER_SHOW_AFTER" = screenshots    \
	     # -o "$LAST_EP_NUMBER_SHOW_AFTER" = both        ]  \
	# && print_last_shown_episode_number


echo
exit 0

# IDEAS
# ————————
# Add ED/OP as something equal to episode numbers in terms of sequences?
# Make a true check for multiple manual rearrangements expression.
# Create journal header to contain version for compatibility check.
# And integration with MyAnimeList. (Would be a pain with cases like Tekyuu (see below))
# Write modular localization.
# Rewrite it with C++.
#
# I’ve given it much thought, but still can’t decide, whether I should reduce heuristics levels down to two,
#   i.e. ‘enabled’ and ‘disabled’. On the one hand, this is how it’s meant to be – user either uses full-featured
#   heuristics or satisfies with basic sort that saves his CPU time. On the other hand, it’s hard to keep up
#   all stages clean from bugs, and redevelop early stages without altering the ones above them. The posiibility
#   to gradually disable applied heuristics methods makes it possible to work on deep stages and see the results.
# Switching between heuristic levels with ‘h’ key would be ideal with two levels, but having three is not much
#   of a burden, considering that it’s meant to be set [to 2] once and for all in an alias, and remembered only
#   when something goes… not as it should. I thought about reducing levels down to two in ‘user-mode’, i.e. make
#   it a part of usual behaviour, enabling three levels when D (for DEBUG) is set, but what if
#   - there are users who’d prefer a bit faster heuristics for the more precise one for the common use?
#   - there are some users who would like to switch heuristics level just because of babyduck syndrome or
#     unexplainable love to click switches (aka hacky and cool)?
#   - there are something more I didn’t think about? Test cases are very small.
#
# There are cases such as Teekyuu when all the episode can be marked sequentially, i.e. each new season continues
#   numbering insteadof starting from 01, 02… etc. Should the script try to find this case, if number of subtitle
#   files matches the number in the first group (simplest case?) But it may work bad in really messed up folders,
#   which would involve HEU2… when it will be finished.

# WAT DONE
# ————————

# TODO
# ————————
#
#   1. Check that in every damn find clause brackets are escaped.
#      Fucking find ignores ? * and lone [ and ] in -name "pattern",
#      but if said pattern will contain both [ and ], find will think
#      it’s time to enable GLOB pattern matching. Re⋅tar⋅da⋅tion.
#
#
#
# Add pstree to deps in deb and rpm.
# The purpose of this update is^W shoud have been to make HEU1 better (actual sorting by number included) and to make HEU2
#   ready for use, so this program would put on title ‘version 1.0’ with dignity.
# GROUP_INDICATOR must be 4 unique chars! Or not? Try to combine an array which would contain indices of groups
#   and start/end indices of the elements in VIDITEM_FILE[@], so we could use them instead of GI_*. // GI_*
#   must be unique in that case, and that may be inconvenient for the user and will require another cycle
#   for checking his own GROUP_INDICATOR.
# Group placed at the top, however, will still persist, and the topmost will be used for complementing in HEU2.
